### ISSUE TITLE
[JENKINS-63169] Jenkins hangs after Listing all plugins

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-21#296]
<p>After restarting Jenkins (cold-load of the system), Jenkins came up once. We took it down cleanly, then restarted and it hung after the message "Listed all plugins". We waited 12 hours for Jenkins to come up without success or additional messages. It usually starts in under 15 seconds from this point.</p>

<p>Thread dump is attached.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsbeckerca.jpg" title="rsbeckerca's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rsbeckerca">rsbeckerca</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63169" target="_blank">Jenkins hangs after Listing all plugins</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Blocker
<li><b>component(s)</b>: core
<li><b>label(s)</b>: itanium
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2025-10-27T22:02:57+00:00
<li><b>votes</b>: 1
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-11-22
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>After restarting Jenkins (cold-load of the system), Jenkins came up once. We took it down cleanly, then restarted and it hung after the message "Listed all plugins". We waited 12 hours for Jenkins to come up without success or additional messages. It usually starts in under 15 seconds from this point.</p>

<p>Thread dump is attached.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins:2.248<br/>
Java:1.8.0_25<br/>
Platform HPE NonStop TNS/E
```
</details>
<details><summary><i>1 attachment</i></summary>

- [jenkins_hang.txt](https://issues.jenkins.io/secure/attachment/51968/jenkins_hang.txt)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsbeckerca.jpg" title="rsbeckerca's avatar" /> rsbeckerca:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63169?focusedId=394772&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-394772">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We were able to bring up jenkins by removing the jobs directory. By hypothesis, there may be something corrupt associated with one of the jobs.</p>
</pre>
</details>

</ul>
</details>

<p>We were able to bring up jenkins by removing the jobs directory. By hypothesis, there may be something corrupt associated with one of the jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63169?focusedId=395079&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-395079">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi. Sorry for the late response. So here is the sequence of events:</p>
<ol>
	<li>There is a onInitMilestoneAttained() handler in the Jenkins core (second thread). It runs after all plugins are prepared. <a href="https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/model/Jenkins.java#L1154-L1165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/model/Jenkins.java#L1154-L1165 </a></li>
	<li>The hook runs the extension loading in hudson.ClassicPluginStrategy.findComponent()</li>
	<li>Extension loading iterates all Extension annotations using the sezpoz library. The library loads classes by finally invoking the native java.lang.Class.forName0() method. There is no evidence of the method being blocked, but it definitely is NOT expected to run for 12 hours</li>
	<li>Project Inheritance Plugin defines the `onJenkinsStart()` startup hook which runs after all jobs are loaded. <a href="https://github.com/i-m-c/jenkins-inheritance-plugin/blob/master/src/main/java/hudson/plugins/project_inheritance/projects/creation/CreationItemListener.java#L94-L95" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/i-m-c/jenkins-inheritance-plugin/blob/master/src/main/java/hudson/plugins/project_inheritance/projects/creation/CreationItemListener.java#L94-L95</a></li>
	<li>This hook initializes the job recreation, and it loads configurations. It is a legitimate behavior</li>
</ol>


<p>Cause analysis:</p>
<ul>
	<li>Reactor listener and the onInitMilestoneAttained() handler are designed to be invoked asynchronously. So execution of the two calls in parallel and temporary locking is an expected behavior. OTOH it is not really usual that they run in parallel from what I remember.  HPE NonStop TNS is an Itanium platform, it may have something to do with that</li>
	<li>Without multiple stacktraces and logs I cannot say for sure whether java.lang.Class.forName0() actually hangs</li>
	<li>Java:1.8.0_25 is reported as a Java version though the stacktrace displays the 25.45-b02 build number. This version is, ehm, very old (2015 or so). 8u261 is the current baseline provided by Oracle</li>
	<li>It is highly likely that one of the recent changes in the core is not fully compatible with such an old JVM. Itanium platform does not help, because we do not test Jenkins on this platform at the moment.</li>
</ul>


<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rsbeckerca" class="user-hover" rel="rsbeckerca">rsbeckerca</a> my recommendation would be to start from updating to a recent Java version. If it does not help, please collect the startup logs and a time series of stacktractes (e.g. every 5 seconds)</p>

<p> </p>

<p> </p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>"CreationItemListener.onJenkinsStart" #26 daemon prio=5 os_prio=24 tid=0x0000000403243000 nid=0x1c waiting for monitor entry [0x00000000656fd000] java.lang.Thread.State: BLOCKED (on object monitor) at hudson.ExtensionList.ensureLoaded(ExtensionList.java:315)
 - waiting to lock &lt;0x00000007267b5890&gt; (a hudson.ExtensionList$Lock) at hudson.ExtensionList.iterator(ExtensionList.java:172)
...
     at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1173)
     at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1053)
 5 &gt;&gt;&gt; at hudson.XmlFile.read(XmlFile.java:149)
     at hudson.plugins.project_inheritance.projects.creation.ProjectCreationEngine.&lt;init&gt;(ProjectCreationEngine.java:380)
     at hudson.plugins.project_inheritance.projects.creation.ProjectCreationEngine.&lt;clinit&gt;(ProjectCreationEngine.java:317)
 4 &gt;&gt;&gt; at hudson.plugins.project_inheritance.projects.creation.CreationItemListener.onJenkinsStart(CreationItemListener.java:119)
     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.lang.reflect.Method.invoke(Method.java:497)
 at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:104)
 at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:175)
 at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:296)
 at jenkins.model.Jenkins$5.runTask(Jenkins.java:1131)
 at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:214)
 at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
 at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
 at java.lang.Thread.run(Thread.java:745)
 
</pre>
</div></div>
<p> </p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>"pool-6-thread-1" #25 daemon prio=5 os_prio=24 tid=0x0000000403251400 nid=0x1b in Object.wait() [0x00000000657fe000]
 java.lang.Thread.State: RUNNABLE
3 &gt;&gt;&gt; at java.lang.Class.forName0(Native Method)
 at java.lang.Class.forName(Class.java:365)
 at hudson.ExtensionFinder$Sezpoz.scout(ExtensionFinder.java:733)
2 &gt;&gt;&gt;  at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:340)
 at hudson.ExtensionList.load(ExtensionList.java:381)
 at hudson.ExtensionList.ensureLoaded(ExtensionList.java:317)
 - locked &lt;0x00000007267b5890&gt; (a hudson.ExtensionList$Lock)
 at hudson.ExtensionList.getComponents(ExtensionList.java:183)
1 &gt;&gt;&gt; at jenkins.model.Jenkins$6.onInitMilestoneAttained(Jenkins.java:1161)
 at jenkins.InitReactorRunner$1.onAttained(InitReactorRunner.java:85)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator.lambda$onAttained$3(ReactorListener.java:102)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator$$Lambda$48/312649756.run(Unknown Source)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator.run(ReactorListener.java:109)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onAttained(ReactorListener.java:102)
 at org.jvnet.hudson.reactor.Reactor$1.run(Reactor.java:177)
 at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
 at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
 at java.lang.Thread.run(Thread.java:745)</pre>
</div></div>
<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi. Sorry for the late response. So here is the sequence of events:</p>
<ol>
	<li>There is a onInitMilestoneAttained() handler in the Jenkins core (second thread). It runs after all plugins are prepared. <a href="https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/model/Jenkins.java#L1154-L1165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/model/Jenkins.java#L1154-L1165 </a></li>
	<li>The hook runs the extension loading in hudson.ClassicPluginStrategy.findComponent()</li>
	<li>Extension loading iterates all Extension annotations using the sezpoz library. The library loads classes by finally invoking the native java.lang.Class.forName0() method. There is no evidence of the method being blocked, but it definitely is NOT expected to run for 12 hours</li>
	<li>Project Inheritance Plugin defines the `onJenkinsStart()` startup hook which runs after all jobs are loaded. <a href="https://github.com/i-m-c/jenkins-inheritance-plugin/blob/master/src/main/java/hudson/plugins/project_inheritance/projects/creation/CreationItemListener.java#L94-L95" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/i-m-c/jenkins-inheritance-plugin/blob/master/src/main/java/hudson/plugins/project_inheritance/projects/creation/CreationItemListener.java#L94-L95</a></li>
	<li>This hook initializes the job recreation, and it loads configurations. It is a legitimate behavior</li>
</ol>


<p>Cause analysis:</p>
<ul>
	<li>Reactor listener and the onInitMilestoneAttained() handler are designed to be invoked asynchronously. So execution of the two calls in parallel and temporary locking is an expected behavior. OTOH it is not really usual that they run in parallel from what I remember.  HPE NonStop TNS is an Itanium platform, it may have something to do with that</li>
	<li>Without multiple stacktraces and logs I cannot say for sure whether java.lang.Class.forName0() actually hangs</li>
	<li>Java:1.8.0_25 is reported as a Java version though the stacktrace displays the 25.45-b02 build number. This version is, ehm, very old (2015 or so). 8u261 is the current baseline provided by Oracle</li>
	<li>It is highly likely that one of the recent changes in the core is not fully compatible with such an old JVM. Itanium platform does not help, because we do not test Jenkins on this platform at the moment.</li>
</ul>


<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rsbeckerca" class="user-hover" rel="rsbeckerca">rsbeckerca</a> my recommendation would be to start from updating to a recent Java version. If it does not help, please collect the startup logs and a time series of stacktractes (e.g. every 5 seconds)</p>

<p> </p>

<p> </p>


```
"CreationItemListener.onJenkinsStart" #26 daemon prio=5 os_prio=24 tid=0x0000000403243000 nid=0x1c waiting for monitor entry [0x00000000656fd000] java.lang.Thread.State: BLOCKED (on object monitor) at hudson.ExtensionList.ensureLoaded(ExtensionList.java:315)
 - waiting to lock <0x00000007267b5890> (a hudson.ExtensionList$Lock) at hudson.ExtensionList.iterator(ExtensionList.java:172)
...
     at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1173)
     at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1053)
 5 >>> at hudson.XmlFile.read(XmlFile.java:149)
     at hudson.plugins.project_inheritance.projects.creation.ProjectCreationEngine.<init>(ProjectCreationEngine.java:380)
     at hudson.plugins.project_inheritance.projects.creation.ProjectCreationEngine.<clinit>(ProjectCreationEngine.java:317)
 4 >>> at hudson.plugins.project_inheritance.projects.creation.CreationItemListener.onJenkinsStart(CreationItemListener.java:119)
     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.lang.reflect.Method.invoke(Method.java:497)
 at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:104)
 at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:175)
 at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:296)
 at jenkins.model.Jenkins$5.runTask(Jenkins.java:1131)
 at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:214)
 at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
 at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
 at java.lang.Thread.run(Thread.java:745)
 
```
<p> </p>


```
"pool-6-thread-1" #25 daemon prio=5 os_prio=24 tid=0x0000000403251400 nid=0x1b in Object.wait() [0x00000000657fe000]
 java.lang.Thread.State: RUNNABLE
3 >>> at java.lang.Class.forName0(Native Method)
 at java.lang.Class.forName(Class.java:365)
 at hudson.ExtensionFinder$Sezpoz.scout(ExtensionFinder.java:733)
2 >>>  at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:340)
 at hudson.ExtensionList.load(ExtensionList.java:381)
 at hudson.ExtensionList.ensureLoaded(ExtensionList.java:317)
 - locked <0x00000007267b5890> (a hudson.ExtensionList$Lock)
 at hudson.ExtensionList.getComponents(ExtensionList.java:183)
1 >>> at jenkins.model.Jenkins$6.onInitMilestoneAttained(Jenkins.java:1161)
 at jenkins.InitReactorRunner$1.onAttained(InitReactorRunner.java:85)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator.lambda$onAttained$3(ReactorListener.java:102)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator$$Lambda$48/312649756.run(Unknown Source)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator.run(ReactorListener.java:109)
 at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onAttained(ReactorListener.java:102)
 at org.jvnet.hudson.reactor.Reactor$1.run(Reactor.java:177)
 at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
 at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
 at java.lang.Thread.run(Thread.java:745)```
<p> </p>

<p> </p>

<!-- ### Imported Jira references for easier searching -->
<!-- [jira_issue_key=JENKINS-63169] -->
<!-- [jira_comment_id=395079] -->
<!-- [comment_author=oleg_nenashev] -->

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsbeckerca.jpg" title="rsbeckerca's avatar" /> rsbeckerca:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63169?focusedId=395090&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-395090">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We found two conditions that were occurring at startup.</p>
<ol>
	<li>JGit was throwing Arithmetic Overflow conditions because the file system does not support nanosecond resolution. The UNDEFINED (Long.MAX_VALUE) value is unchecked, and any addition with it causes the overflow. This was separately reported to the JGit team.</li>
	<li>The CACERTS were out of date for the jenkins update site. So none of the connections done at startup were working. These went without logging exceptions but fixing the CACERTS allowed Jenkins to start up correctly.</li>
</ol>


<p>As far as upgrading the JVM, this is not something within our team's control. The vendor (HPE) is not releasing new JVM versions on the Itanium platform although we have one more possible upgrade. I will let the team know about this restriction.</p>
</pre>
</details>

</ul>
</details>

<p>We found two conditions that were occurring at startup.</p>
<ol>
	<li>JGit was throwing Arithmetic Overflow conditions because the file system does not support nanosecond resolution. The UNDEFINED (Long.MAX_VALUE) value is unchecked, and any addition with it causes the overflow. This was separately reported to the JGit team.</li>
	<li>The CACERTS were out of date for the jenkins update site. So none of the connections done at startup were working. These went without logging exceptions but fixing the CACERTS allowed Jenkins to start up correctly.</li>
</ol>


<p>As far as upgrading the JVM, this is not something within our team's control. The vendor (HPE) is not releasing new JVM versions on the Itanium platform although we have one more possible upgrade. I will let the team know about this restriction.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63169?focusedId=457219&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-457219">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Closing as "Won't fix" since Jenkins no longer supports Java 8 or Java 11.</p>
</pre>
</details>

</ul>
</details>

<p>Closing as "Won't fix" since Jenkins no longer supports Java 8 or Java 11.</p>
